lib/cosmos/tools/cmd_tlm_server/api.rb

```diff
@@ -670,17 +673,26 @@ def normalize_tlm(*args)
     private
 
     def _override(method, tgt_pkt_item)
-      interface = System.targets[tgt_pkt_item[0]].interface
-      found = false
-      interface.read_protocols.each do |protocol|
-        found = true if protocol.kind_of? OverrideProtocol
-      end
-      if found
-        # Test to see if this telemetry item exists
-        System.telemetry.value(tgt_pkt_item[0], tgt_pkt_item[1], tgt_pkt_item[2], :RAW)
-        interface.public_send("_#{method}", *tgt_pkt_item)
-      else
-        raise "Interface #{interface.name} does not have override ability. Is 'PROTOCOL OverrideProtocol' under the interface definition?"
+      # synchronize this method due to the interface.public_send which can cause
+      # ConcurrencyError exceptions in JRuby
+      @mutex.synchronize do
```

Comment:
I'm not exactly sure why the mutex synchronization was necessary in this case so my comment may not be correct. JRuby definitely didn't like it without the mutex synchronization through.

Reply:
So the error was:

     ConcurrencyError:
       Detected invalid array contents due to unsynchronized modifications with concurrent users
     # ./lib/cosmos/tools/cmd_tlm_server/api.rb:688:in `_override'
     # ./lib/cosmos/tools/cmd_tlm_server/api.rb:654:in `override_tlm_raw'
     # ./spec/tools/cmd_tlm_server/api_spec.rb:841:in `block in Cosmos'

In the spec, it's not like there are two threads both trying to call override at the same time.  I can see how that the underlying @override_tlm array (in interface.rb) could conceivably be concurrently accessed by the test thread and an interface thread, if an interface was calling read().  However, I don't understand how the mutex would solve that.  The mutex is only keeping concurrent threads from calling _override, but it seems like the problematic access of the @override_tlm array would be in override_protocol.rb, which is using the accessor.  That accessor doesn't acquire the mutex, so it seems like concurrent access should still be possible.

I don't quite get this yet... I'm worried that the synchronize might have just changed the timing without actually protecting against concurrent access.